tensorflow_probability/python/experimental/mcmc/sample_fold.py

```diff
@@ -114,5 +123,9 @@ def sample_fold(
         reducers,
         final_kernel_results.streaming_calculations,
         check_types=False)
-    warm_restart_package = (end_state, final_kernel_results)
-    return reduction_results, warm_restart_package
+    if reduction_results == []:
+      reduction_results = None
```

Comment:
So, this doesn't actually work either, because if the user had passed `reducers=[]` explicitly, they will get `None` reduction results, and should get `[]`.  That's why input canonicalization mixes poorly with maintaining the user's choice of data structures: canonicalization loses information that must then be re-constructed.

In this case, you have argued that the underlying tf.while_loop can't handle a `None` reducers structure.  If that is the case, then something has to deal with it, like this:
```
if reducers is None:
  reducers_was_none = True
  reducers = []
...
if reducers_was_none:
  reducer_parallel_results = None
```

If it doesn't make sense to put this into `WithReductions`, then it unfortunately has to go here.